README.md
# Ardustat

Схема в Fritzing:

![Fritzing](./readme_assets/fritzing.png)

Термостат на Arduino Nano + NTC с 4-разрядным 7-сегментным дисплеем.

Код и схема создавались в онлайн-симуляторе [Wokwi](https://wokwi.com/) и не проверялись в реальности.

Настройки сохраняются в EEPROM. Беспокоиться не стоит, используется библиотека EEManager с удобной логикой.

Занятые ресурсы МК по мнению PlatformIO:

```
RAM:   [=         ]  10.0% (used 205 bytes from 2048 bytes)
Flash: [==        ]  24.7% (used 7586 bytes from 30720 bytes)
```

## Зависимости - в [platformio.ini](./platformio.ini)

## Электронные компоненты

- Arduino Nano
- Реле на 5V на ток `мощность / напряжение` + запас. Например ТЭН на 2КВт с напряжением 220V потребляет ток ~9A - лучше взять реле с запасом, например на 16 ампер. На али лично я не видел модулей реле выше 10 ампер, поэтому можно взять твердотельное. Но оно существенно дороже обычного - 500 рублей против 60. Зато оно не боится частых переключений, не нужна никакая обвязка, контакты не могут износиться (их там нет -_-). Минус твердотельного реле - оно греется, поэтому нужно брать с запасом по току и/или крепить на радиатор
- Модуль энкодера с кнопкой с обвязкой против дребезга ([инфо](https://github.com/GyverLibs/EncButton#%D1%80%D0%B5%D0%BA%D0%BE%D0%BC%D0%B5%D0%BD%D0%B4%D0%B0%D1%86%D0%B8%D0%B8)). На али у него круглая синяя плата
- Термистор
- Резистор, равный сопротивлению термистора
- Дисплей TM1637

## Руководство пользователя

В формате .Docx - в нем удобнее делать дизайн для печати.

Приложено к Github релизу и [лежит тут](./USER_MANUAL.docx)

## А как прошить, а как собрать

- [Уроки Arduino](https://alexgyver.ru/arduino-first/)
- [Оф гайд Platformio CLI](https://docs.platformio.org/en/latest/core/quickstart.html)
- [Оф гайд Platformio IDE (VSCode)](https://docs.platformio.org/en/latest/integration/ide/vscode.html)

## История версий - в [CHANGELOG.md](./CHANGELOG.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog],
and this project adheres to [Semantic Versioning].

## [1.1.1] 2023-12-10

### Added

- Added prefix `AS_` (Ardustat => ArduStat => AS) for all `#define`

### Removed

- Removed useless settings in `config.h`
- Removed debug messages
- Removed the `F()` macro - useless for strings a couple of characters
- Removed unnecessary mode check when re-rendering the display after reading the temperature

### Changed

- Renamed some variables to reduce code. Example - `readed_temp` => `sens_temp`, `hysteresis` => `hstr`
- Merged functions `print_mode` and `put_num_at_end`
- Simplified method for checking message type (message or error)
- Simplified encoder handling function

## [1.1.0] 2023-12-10

### Added

- Added `#pragma once` in all headers
- Added [macro `F()`](https://alexgyver.ru/lessons/code-optimisation/) in all static strings

### Removed

- Removed endless error display

### Changed

- Merged Error and Message modes
- Mode symbols are included in the config
- Includes are placed in main.h
- Increased the number of comments to the code

## [1.0.1] 2023-12-09

### Changed

- Switch to PlatformIO

## [1.0.0] - 2023-12-09

- First version


src/main.cpp
#include "main.h"

// Переменные
int16_t sens_temp = AS_DEFAULT_SENS_TEMP;
bool need_redraw_display = false;
uint8_t mode = AS_MODE_DEFAULT;
uint8_t message_code = AS_MESSAGE_NO;
bool message_is_error = false;

// Изменяемые настройки
struct Settings {
  int16_t temp = AS_DEFAULT_TEMP;
  int16_t hstr = AS_DEFAULT_HSTR;
};

Settings settings;

// Инициализация объектов
EEManager memory(settings, AS_SETTINGS_SAVE_TIME);
EncButton enc(AS_PIN_ENC_A, AS_PIN_ENC_B, AS_PIN_ENC_BUTT);
Disp1637Colon display(AS_PIN_TM1637_DIO, AS_PIN_TM1637_CLK);
GyverNTC ntc(AS_PIN_NTC, AS_NTC_RESISTOR, AS_NTC_BETA_COEF, AS_NTC_BASE_TEMP,
             AS_NTC_BASE_RESISTANCE, AS_NTC_RESOLUTION);
TimerMs temp_read_timer(AS_TEMP_READ_TIME, 1, 0);
TimerMs relay_update_timer(AS_RELAY_UPDATE_TIME, 1, 0);
TimerMs message_close_timer(AS_MESSAGE_TIME, 0, 1);

// Планируем отобразить сообщение при следующей отрисовке дисплея
void add_message() {
  mode = AS_MODE_MESSAGE;

  message_close_timer.start();

  need_redraw_display = true;
}

// Планируем убрать сообщение при следующей отрисовке дисплея
void remove_message() {
  mode = AS_MODE_DEFAULT;

  need_redraw_display = true;
}

// Читаем температуру
void read_temp() { sens_temp = ntc.getTempAverage(); }

// Обновляем состояние реле
void update_relay() {
  // Гистерезис - https://alexgyver.ru/lessons/relay-control/
  if (sens_temp < (settings.temp - settings.hstr))
    digitalWrite(AS_PIN_RELAY, HIGH);
  else if (sens_temp > (settings.temp + settings.hstr))
    digitalWrite(AS_PIN_RELAY, LOW);
}

// Меняем режим
void change_mode() {
  mode++;
  if (mode > AS_MODE_SWITCH_MAX) mode = AS_MODE_SWITCH_MIN;

  need_redraw_display = true;
}

// Печатаем режим (его символ и число) (но не обновляем дисплей!)
void print_mode(const char *mode_symbol, int16_t num) {
  display.print(mode_symbol);

  // Печатаем число в конце экрана
  uint8_t num_len = sseg::intLen(num);
  display.setCursor(AS_TM1637_DIGITS_NUM - num_len);
  display.print(num);
}

// Обновление дисплея
void redraw_display() {
  display.setCursor(0);
  display.clear();

  if (mode == AS_MODE_TEMP)
    print_mode(AS_MODE_SYMBOL_TEMP, settings.temp);
  else if (mode == AS_MODE_SENS_TEMP)
    print_mode(AS_MODE_SYMBOL_SENS_TEMP, sens_temp);
  else if (mode == AS_MODE_HSTR)
    print_mode(AS_MODE_SYMBOL_HSTR, settings.hstr);
  else if (mode == AS_MODE_MESSAGE) {
    print_mode(message_is_error ? AS_MODE_SYMBOL_ERROR : AS_MODE_SYMBOL_MESSAGE,
               message_code);

    // Вывели - чистим за собой
    message_code = AS_MESSAGE_NO;
    message_is_error = false;
  } else {
    // Если ID режима не подходит, ставим ошибку. При следующем рендере она
    // покажется
    message_code = AS_MESSAGE_ERROR_DISPLAY;
    message_is_error = true;
  }

  display.update();

  need_redraw_display = false;
}

// Работа с энкодером
void enc_handle() {
  enc.tick();

  if (enc.click()) change_mode();

  if (enc.turn()) {
    // `-1` - шаг изменения, указывается отрицательным числом
    if (mode == AS_MODE_TEMP)
      settings.temp =
          constrain(settings.temp + -1 * enc.dir(), AS_MIN_TEMP, AS_MAX_TEMP);
    else if (mode == AS_MODE_HSTR)
      settings.hstr =
          constrain(settings.hstr + -1 * enc.dir(), AS_MIN_HSTR, AS_MAX_HSTR);

    memory.update();
    need_redraw_display = true;
  }
}

// Функция обработки программного прерывания для энкодера. Читайте
// https://github.com/GyverLibs/EncButton
void enc_isr() { enc.tickISR(); }

void setup() {
#ifdef ENABLE_DEBUG
  Serial.begin(115200);
#endif

  attachPCINT(digitalPinToPCINT(AS_PIN_ENC_A), enc_isr, CHANGE);
  attachPCINT(digitalPinToPCINT(AS_PIN_ENC_B), enc_isr, CHANGE);
  attachPCINT(digitalPinToPCINT(AS_PIN_ENC_BUTT), enc_isr, CHANGE);

  memory.begin(AS_EEPROM_SETTINGS_ADDR, AS_EEPROM_INIT_MARKER_VALUE);

  need_redraw_display = true;
}

void loop() {
  // Если настройки были записаны в EEPROM - ставим радостное сообщение
  if (memory.tick()) message_code = AS_MESSAGE_SAVED;

  if (temp_read_timer.tick()) {
    read_temp();
    need_redraw_display = true;
  }

  enc_handle();

  if (relay_update_timer.tick()) update_relay();

  if (message_code) add_message();
  if (message_close_timer.tick()) remove_message();
  if (need_redraw_display) redraw_display();
}


src/config.h
#pragma once

// Раскомментируй для отладки по Serial
// #define ENABLE_DEBUG

#ifdef ENABLE_DEBUG
#define DEBUG(x) Serial.print(x)
#define DEBUGLN(x) Serial.println(x)
#else
#define DEBUG(x)
#define DEBUGLN(x)
#endif

// ======== Пины

// Пин термистора
#define AS_PIN_NTC A0
// Пин реле
#define AS_PIN_RELAY 2
// Пин энкодера A
#define AS_PIN_ENC_A 5
// Пин энкодера B
#define AS_PIN_ENC_B 4
// Пин кнопки энкодера
#define AS_PIN_ENC_BUTT 3
// Пин CLK дисплея TM1637
#define AS_PIN_TM1637_CLK 11
// Пин DIO дисплея TM1637
#define AS_PIN_TM1637_DIO 12

// ======== Разные настройки

// Минимальная температура
#define AS_MIN_TEMP 20
// Максимальная температура
#define AS_MAX_TEMP 60
// Температура по умолчанию
#define AS_DEFAULT_TEMP 20
// Температура с датчика по умолчанию
#define AS_DEFAULT_SENS_TEMP 99
// Мин. гистерезис
#define AS_MIN_HSTR 1
// Макс. гистерезис
#define AS_MAX_HSTR 10
// Гистерезис по умолчанию
#define AS_DEFAULT_HSTR 2

// ======== Таймауты (в миллисекундах)

// Таймаут замера температура
#define AS_TEMP_READ_TIME 500
// Время между переключениями реле
#define AS_RELAY_UPDATE_TIME 3000
// Время показа сообщения
#define AS_MESSAGE_TIME 3000
// Задержка сохранения настроек
#define AS_SETTINGS_SAVE_TIME 5000

// ======== Настройки EEPROM

#define AS_EEPROM_INIT_MARKER_VALUE \
  10  // Значение флага по которому устройство будет понимать что уже включалось
      // и в нем что-то сохранено. Взято от балды, но не должно быть
      // равно стандартному значению в EEPROM `255`

#define AS_EEPROM_SETTINGS_ADDR 0  // Адрес с настройками

// ======== Настройки дисплея

// Количество разрядов на дисплее
#define AS_TM1637_DIGITS_NUM 4

// ======== Режимы

// Заданная температура
#define AS_MODE_TEMP 0
// Текущая температура
#define AS_MODE_SENS_TEMP 1
// Гистерезис
#define AS_MODE_HSTR 2
// Сообщение
#define AS_MODE_MESSAGE 101
// Стандартный режим
#define AS_MODE_DEFAULT AS_MODE_TEMP
// Первый режим доступный для открытия
#define AS_MODE_SWITCH_MIN 0
// Последний режим доступный для открытия
#define AS_MODE_SWITCH_MAX 2

// ======== Символы режимов

#define AS_MODE_SYMBOL_TEMP "S"
#define AS_MODE_SYMBOL_SENS_TEMP "C"
#define AS_MODE_SYMBOL_HSTR "H"
#define AS_MODE_SYMBOL_ERROR "E"
#define AS_MODE_SYMBOL_MESSAGE "I"

// ======== Коды сообщений

#define AS_MESSAGE_NO 0
#define AS_MESSAGE_SAVED 1
#define AS_MESSAGE_ERROR_DISPLAY 101

// ======== NTC (термистор)

// Описание с сайта AlexGyver:
// https://kit.alexgyver.ru/tutorials/thermistor
//
// - AS_NTC_RESISTOR - сопротивление балластного резистора, Ом
// - AS_NTC_BETA_COEF - бета-коэффициент термистора (см. даташит) [число в
// районе 1000-5000]
// - AS_NTC_BASE_TEMP - базовая температура термистора, градусов Цельсия (см.
// даташит) [обычно 25 градусов]
// - AS_NTC_BASE_RESISTANCE - сопротивление термистора при базовой температуре,
// Ом (см. даташит)
// - AS_NTC_RESOLUTION - разрешение АЦП, бит. По умолчанию 10

#define AS_NTC_RESISTOR 10000
#define AS_NTC_BETA_COEF 3950
#define AS_NTC_BASE_TEMP 25
#define AS_NTC_BASE_RESISTANCE 10000
#define AS_NTC_RESOLUTION 10
